include/cuco/detail/static_map.inl

```diff
@@ -164,23 +217,105 @@ __device__ bool static_map<Key, Value, Scope, Allocator>::device_mutable_view::i
+static_map<Key, Value, Scope, Allocator>::device_mutable_view::insert(CG g,
+                                                                      value_type const& insert_pair,
+                                                                      Hash hash,
+                                                                      KeyEqual key_equal) noexcept
 {
+  using cuda::std::memory_order_relaxed;
   auto current_slot = initial_slot(g, insert_pair.first, hash);
 
   while (true) {
-    key_type const existing_key = current_slot->first.load(cuda::memory_order_relaxed);
+    key_type const existing_key = current_slot->first.load(memory_order_relaxed);
 
     // The user provide `key_equal` can never be used to compare against `empty_key_sentinel` as the
     // sentinel is not a valid key value. Therefore, first check for the sentinel
-    auto const slot_is_empty = detail::bitwise_compare(existing_key,this->get_empty_key_sentinel());
+    auto const slot_is_empty =
+      detail::bitwise_compare(existing_key, this->get_empty_key_sentinel());
 
     // the key we are trying to insert is already in the map, so we return with failure to insert
-    if (g.ballot(not slot_is_empty and key_equal(existing_key, insert_pair.first))) {
-      return false;
+    if (g.any(not slot_is_empty and key_equal(existing_key, insert_pair.first))) { return false; }
+
+    auto const window_contains_empty = g.ballot(slot_is_empty);
+
+    // we found an empty slot, but not the key we are inserting, so this must
+    // be an empty slot into which we can insert the key
+    if (window_contains_empty) {
+      // the first lane in the group with an empty slot will attempt the insert
+      insert_result status{insert_result::CONTINUE};
+      uint32_t src_lane = __ffs(window_contains_empty) - 1;
+
+      if (g.thread_rank() == src_lane) {
+        auto expected_key   = this->get_empty_key_sentinel();
+        auto expected_value = this->get_empty_value_sentinel();
+
+        cuco::detail::pair_converter<pair_type> converter;
+
+        converter.pair =
+          cuco::make_pair<Key, Value>(std::move(expected_key), std::move(expected_value));
+        auto empty_sentinel = converter.packed;
+        converter.pair      = insert_pair;
+        auto tmp_pair       = converter.packed;
+
+        auto slot = reinterpret_cast<
+          cuda::atomic<typename cuco::detail::pair_converter<pair_type>::packed_type>*>(
+          current_slot);
+
+        bool success =
+          slot->compare_exchange_strong(empty_sentinel, tmp_pair, memory_order_relaxed);
+        converter.packed   = empty_sentinel;
+        auto expected_pair = converter.pair;
+        if (success) {
+          status = insert_result::SUCCESS;
+        }
+        // duplicate present during insert
+        else if (key_equal(insert_pair.first, expected_pair.first)) {
+          status = insert_result::DUPLICATE;
+        }
+      }
```

Comment:
Is there a way we can reduce the redundancy here with the non-CG insert code?

Reply:
Any suggestions?

Reply:
"Zero" redundancy right now.